# Helper Scripts

:warning: These scripts are _experimental_. They may be classified as 'hacky' and have not been thoroughly debugged and tested. :warning:

This directory holds the optional scripts for setup.

## Downloading Reads

Reads can be downloaded using `get-reads.sh`, where `runs.txt` is a text file with one SRR acccession per line. An SRA accession list (text file, 1 per line) can be converted into an SRR accession list using `get-runs.sh`.

### _P. toftae_ example

`accessions.txt`, in `amphibia/ptoftae/skin-liver`:

```
SRX5102741-46
SRX5102761-62
```

To download those reads, the first step is to fetch the SRR accessions and metadata corresponding to the _SRA_ accessions:

```
USAGE(S):
      get-runs.sh [OPTIONS] -o <output directory> <SRA accessions TXT file>
```

```shell
scripts/helpers/get-runs.sh -o amphibia/ptoftae/skin-liver/sra amphibia/ptoftae/skin-liver/accessions.txt
```

Next, `get-reads.sh` is used to download the reads using `amphibia/ptoftae/skin-liver/sra/runs.txt` (output from last step), as input:

```
USAGE(S):
      get-reads.sh [OPTIONS] -o <output directory> <SRA RUN (i.e. SRR) accession list>
```

```shell
scripts/helpers/get-reads.sh -o amphibia/ptoftae/skin-liver/raw_reads amphibia/ptoftae/skin-liver/sra/runs.txt
```

## Making A Reads List

`make-reads-list.sh` can be used to 'automatically' make a reads list using the `metadata.tsv` file generated by `get-runs.sh`. This script attempts to choose a "pooling" factor based on the similarities and differences of each column in the metadata.

:warning: These reads list should be double checked as the inference of pooling factor is not always correct. :warning:

**Note**: the environment variables `PAIRED` and `STRANDED` need to have been exported before using `make-reads-list.sh`.

```
USAGE(S):
      make-reads-list.sh [OPTIONS] -d <I/O directory> <metadata TSV file>
```

### _P. toftae_ example

Since the _P. toftae_ dataset has non-stranded paired-end reads, we export those variables before making the reads list:

```shell
export STRANDED=false
export PAIRED=true
scripts/helpers/make-reads-list.sh -d amphibia/ptoftae/skin-liver/trimmed_reads amphibia/ptofate/skin-liver/sra/metadata.tsv
```

## Downloading the AMP Databases

AMP sequences have been provided in `amp_seqs/` and do not need to be downloaded unless they need to be updated.

```
USAGE(S):
      homology-db.sh [OPTIONS] -o <output directory>
```

```shell
scripts/helpers/homology-db.sh -o amp_seqs
```

## Downloading Reference Transcriptomes

`dl-ref.sh` can be used to download reference transcriptomes.

```
USAGE(S):
      dl-ref.sh [OPTIONS] -o <output directory> <FTP Genbank path>
```

### _M. gulosa_ example

```shell
scripts/helpers/dl-ref.sh -o insecta/mgulosa/venom /genbank/tsa/G/tsa.GGFG.1.fsa_nt.gz
```

## Configurations

### Configuring SignalP

`config-signalp.sh` can be used to configure SignalP.

```
USAGE(S):
        config-signalp.sh [OPTIONS]
```

```shell
scripts/helpers/config-signalp.sh
```

### Configuring ProP

`config-prop.sh` can be used to configure ProP.

```
USAGE(S):
        config-prop.sh [OPTIONS]
```

```shell
scripts/helpers/config-prop.sh
```

### Configuring SABLE

`config-sable.sh` can be used to configure SABLE.

```
USAGE(S):
      config-sable.sh [OPTIONS]
```

```shell
scripts/helpers/config-sable.sh
```