# Helper Scripts

:warning: These scripts are _experimental_. They may be classified as 'hacky' and have not been thoroughly debugged and tested. :warning:

This directory holds the optional scripts for setup.

## Downloading Reads

Reads can be downloaded using `get-reads.sh`, where `runs.txt` is a text file with one SRR acccession per line. An SRA accession list (text file, 1 per line) can be converted into an SRR accession list using `get-runs.sh`.

### `get-runs.sh`

```
PROGRAM: get-runs.sh

DESCRIPTION:
       Gets the SRA RUN (i.e. SRR) accessions using wget.
       
       OUTPUT:
       -------
         - runs.txt
         - metadata.tsv
         - RUNS.DONE
         - METADATA.DONE
       
       
       EXIT CODES:
       -------------
         - 0: successfully completed
         - 1: general error
       
USAGE(S):
      get-runs.sh [OPTIONS] -o <output directory> <SRA accessions TXT file>
      
OPTION(S):
       -a <address>    email alert          
       -h              show this help menu  
       -o <directory>  output directory     (required)
                                            
EXAMPLE(S):
      get-runs.sh -o /path/to/sra /path/to/accessions.txt
```

### `get-reads.sh`

```
PROGRAM: get-reads.sh

DESCRIPTION:
       Gets reads for one single organism, using fasterq-dump.
       
       OUTPUT:
       -------
         - *.fastq.gz
         - RUNS.DONE or RUNS.FAIL
       
       EXIT CODES:
       -------------
         - 0: successfully completed
         - 1: general error
         - 2: failed to download
       
       For more information: https://github.com/ncbi/sra-tools/wiki/HowTo:-fasterq-dump
       
USAGE(S):
      get-reads.sh [OPTIONS] -o <output directory> <SRA RUN (i.e. SRR) accession list>
      
OPTION(S):
       -a <address>    email alert                    
       -h              show this help menu            
       -o <directory>  output directory               (required)
       -p              download each run in parallel  
       -t <int>        number of threads              (default = 2)
                                                      
EXAMPLE(S):
      get-reads.sh -o /path/to/raw_reads /path/to/sra/runs.txt
```

## Expanding Accessions

`expand-accessions.sh` can be used to expand shorthand accessions like `SRX12345-67` to `SRX12345 SRX12346 SRX12347 SRX12348 SRX12349 SRX12350 SRX12351 SRX12352 SRX12353 SRX12354 SRX12355 SRX12356 SRX12357 SRX12358 SRX12359 SRX12360 SRX12361 SRX12362 SRX12363 SRX12364 SRX12365 SRX12366 SRX12367`.

```
DESCRIPTION:
      Expands SRA accessions that are written with ranges (using hyphens), as individual accessions.
      
USAGE(S):
      expand-accessions.sh <SRA accessions>
      
OPTION(S):
       -h              show this help menu  
       -s <separator>  output separator     (default = space)
                                            
EXAMPLE(S):
      expand-accessions.sh SRX12345-48 SRR12350
```

## Making A Reads List

`make-reads-list.sh` can be used to make a reads list using the `metadata.tsv` file generated by `get-runs.sh`. This script attempts to choose a "pooling" factor based on the similarities and differences of each column in the metadata. This reads list should be double checked as it is not always entirely correct.

```
PROGRAM: make-reads-list.sh

DESCRIPTION:
       Makes the pooled reads lists for RNA-Bloom. Filters given TSV for relevant information.
       
       OUTPUT:
       -------
         - reads.txt
         - READSLIST.DONE
       
       EXIT CODES:
       -----------
         - 0: successfully completed
         - 1: general error
       
       For more information: https://github.com/bcgsc/RNA-Bloom
       
USAGE(S):
      make-reads-list.sh [OPTIONS] -d <I/O directory> <metadata TSV file>
      
OPTION(S):
       -a <address>    email alert                                                          
       -d <directory>  Input directory (trimmed reads) and output directory for reads list  (required)
       -h              Show this help menu                                                  
                                                                                            
EXAMPLE(S):
      make-reads-list.sh -d /path/to/trimmed_reads /path/to/sra/metadata.tsv
```

## Downloading the AMP Databases

```
DESCRIPTION:
       Downloads AMP sequences from NCBI protein database: antimicrobial[All Fields] AND class[Organism]
       Downloads AMP sequences from The Antimicrobial Peptide Database (APD3): http://aps.unmc.edu/AP/main.php
       Downloads Anuran AMP sequences from Database of Anuran Defense Peptides (DADP): http://split4.pmfst.hr/dadp/
       
USAGE(S):
      homology-db.sh [OPTIONS] -o <output directory>
      
OPTION(S):
       -h              show this help menu  
       -o <directory>  output directory     (required)
```

## Downloading Reference Transcriptomes

`dl-ref.sh` can be used to download reference transcriptomes.

```
DESCRIPTION:
      Using cURL, gets draft assemblies given accession numbers.
      
USAGE(S):
      dl-ref.sh [OPTIONS] -o <output directory> <FTP Genbank path>
      
OPTION(S):
       -h  show this help menu  
       -o  output directory     (required)
```

## Configurations

### Configuring SignalP

`config-signalp.sh` can be used to configure SignalP.

```
DESCRIPTION:
         Configures the environment variables needed for running SignalP. Uses 'sed' to change variables in /projects/amp/peptaid/src/signalp-3.0/signalp (changes can be made manually as well).
         See specifics below for more information.
         
USAGE(S):
        config-signalp.sh [OPTIONS]
        
OPTION(S):
        -h  show this help menu
            
SPECIFICS:
         SIGNALP=/projects/amp/peptaid/src/signalp-3.0
         SH=/bin/bash
         AWK=/gsc/btl/linuxbrew/bin/awk
```

### Configuring ProP

`config-prop.sh` can be used to configure ProP.

```
DESCRIPTION:
         Configures the environment variables needed for running ProP. Uses 'sed' to change variables in /projects/amp/peptaid/src/prop-1.0c/prop (changes can be made manually as well).
         See specifics below for more information.
         
USAGE(S):
        config-prop.sh [OPTIONS]
        
OPTION(S):
        -h  show this help menu
            
SPECIFICS:
         setenv PROPHOME=/projects/amp/peptaid/src/prop-1.0c
         setenv AWK=/gsc/btl/linuxbrew/bin/awk 
         setenv ECHO=echo -e
         setenv GNUPLOT=/gsc/btl/linuxbrew/bin/gnuplot
         setenv PPM2GIF=/gsc/btl/linuxbrew/bin/ppmtogif
```

### Configuring SABLE

`config-sable.sh` can be used to configure SABLE.

```
DESCRIPTION:
      Configures environment variables for the run.sable script.
      For more information: /projects/amp/peptaid/src/sable_v4_distr/README

USAGE(S):
      config-sable.sh [OPTIONS]

OPTION(S):
      -h  show this help menu  
      -t  number of threads    (default = 8)
```